workflow/docs/design.md

```diff
@@ -0,0 +1,406 @@
+- `SerialScheduler`, it should pick the next **one** template from children templates in SerialTemplate.
+- `ParallelScheduler`, it should pick the all templates from children templates in ParallelTemplate.
+- `TaskScheduler`, it is based on `ParallelScheduler`, but contains logic about evaluation expressions for conditional branches.
+
+```go
+type Scheduler interface{
+  ScheduleNext(ctx context.Context) (nextTemplates []template.Template, parentNodeName string, err error)
+}
+```
+
+### Actor and Playground
+
+> - Nothing happens until the Actor works.
+> - The status of Workflow does NOT present the result for the Actor. (It means we could not fetch the result of one "Actor" from kubernetes.) Because the "real world" is so complicated to describe, we want the result of "Actor" do not "pollution" our schedule objects.
+> - While an "Actor" failed, the whole workflow instance is trapped into failed.
+
+The Actor will do all "dirty" effects, for example: creating Chaos Resource, running user-defined tasks, and so on.
+
+```go
+type Actor interface {
+  PlayOn(pg Playground) error
+}
+
+type Playground interface {
+  CreateNetworkChaos(networkChaos chaosmeshv1alph1.NetworkChaos) error
+  DeleteNetworkChaos(namespace, name string) error
+  CreatePodChaos(podChaos chaosmeshv1alph1.PodChaos) error
+  DeletePodChaos(namespace, name string) error
+
+  // Other ad-hoc methods
+}
+```
+
+### WorkflowManager: rest glue codes
+
+We need a manager who makes these components work together, and controls the main-loop, and update the status of the workflow, and max-concurrency, and so-on.
+
+The main loop looks like:
+
+```go
+func loop(){
+  for {
+    workflowKey, err := trigger.AcquireNext()
+    scheduler, err := fetchSchedulerFor(workflowKey)
+
+    go func(){
+      templates, err := scheduler.scheduleNext()
+      for item := range templates {
+        go func (){
+          // TODO: update node status
+          actor := fetchActor(item)
+          err := actor.Apply()
+          // TODO: update node status
+          if item.NeedReQueue() {
+            // TODO: manually ask for next trigger
+          }
+        }()
+      }
+    }
+  }
+}
+```
+
+### States for Node
+
+There are 8 available phase of one Node:
+
+- Init
+- WaitingForSchedule
+- WaitingForChild
+- Running
+- Evaluating
+- Holding
+- Succeed
+- Failed
+
+**Init** means is the default phase when **Node** just created, means this node is just created, did not effect real world yet.
+
+**WaitingForSchedule** is only available on Serial, Parallel and Task, it means this Node is idle and safe for next scheduling; It is presents:
+
+- For Chaos, Suspend, Task, Parallel:
+  - This Node did not make "Effect" yet.
+- For Serial:
+  - This Node did not create child node yet;
+  - Or previous child node just succeed;
+
+**WaitingForChild** is only available on Serial, Parallel and Task, it means at least 1 child node is in **Running** state.
```

Comment:
What if the child node is `Holding`?

Reply:
`WaitingForChild` also works for children nodes `Holding`. I will update the doc.

Reply:
fixed, PTAL